File: 09.22/프로그래머스/조수연/거리두기 확인하기.kt
package com.ssafy.algorithm

import java.util.*
import kotlin.collections.ArrayList

fun main() {
    val sl = Solution()
    val places = arrayOf(
        arrayOf("POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"),
        arrayOf("POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"),
        arrayOf("PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"),
        arrayOf("OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"),
        arrayOf("PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP")
    )
    println(sl.solution(places))
}

data class Point(val x: Int, val y: Int)

class Solution {
    lateinit var visited : BooleanArray
    lateinit var points : ArrayList<Point>
    lateinit var copy : Array<CharArray>
    fun solution(places: Array<Array<String>>): IntArray {

        var answer = IntArray(5)
        var index = 0

        for (place in places) {
            copy = Array(5){CharArray(5)}
            points = ArrayList()
            var result = 1

            for (i in place.indices) {
                for (j in place[i].indices) {
                    copy[i][j] = place[i][j]
                    if (place[i][j] == 'P') points.add(Point(i,j))
                }
            }

            visited = BooleanArray(points.size)
            var idx = 0

            for (i in place.indices) {
                for (j in place[i].indices) {
                    if (place[i][j] == 'P') {
                        visited[idx] = true
                        idx++
                        if (!check(i,j)){
                            result = 0
                            break
                        }
                    }
                }
                if (result == 0){
                    break
                }
            }

            answer[index] = result
            index++
        }
        return answer
    }

    fun check(row : Int, col : Int) : Boolean{ //맨허튼 거리로 거리두기 지키는지 확인
        for (i in points.indices){
            if (!visited[i]){
                val distance = Math.abs(row - points[i].x) + Math.abs(col - points[i].y) //맨허튼 거리
                if (distance == 1) return false
                else if (distance == 2){
                    if (!isBlcok(row,col,points[i].x,points[i].y)) return false
                }
            }
        }
        return true
    }

    fun isBlcok(row1 : Int, col1 : Int, row2 : Int, col2 : Int) : Boolean { //파티션으로 막혀있는지 판단
        if (row1 - row2 < 0 && col1 - col2 < 0){ //기준에서 P가 오른쪽 아래에 있을 때
            if (copy[row1+1][col1] != 'X' || copy[row1][col1+1] != 'X') return false
        }else if(row1 - row2 < 0 && col1 - col2 > 0) { //기준에서 P가 왼쪽 아래에 있을 때
            if (copy[row1+1][col1] != 'X' || copy[row1][col1-1] != 'X') return false
        }else if(row1 - row2 < 0 && col1 -  col2 == 0){ //기준에서 P가 아래쪽에 있을 때
            if (copy[row1+1][col1] != 'X') return false
        }else if(row1 - row2 == 0 && col1 - col2 < 0) { //기준에서 P가 오른쪽에 있을 때
            if (copy[row1][col1+1] != 'X') return false
        }
        return true
    }
}

File: 2022/09.22/프로그래머스/이아름/거리두기 확인하기.cpp
#include <string>
#include <vector>

using namespace std;
int dx[] = { 1,0,-1,0 };
int dy[] = { 0,1,0,1 };
int N = 5;

bool check(vector<string> &place) {
	for (int i = 0; i < place.size(); i++) {
		for (int k = 0; k < place[i].size(); k++) {
			if (place[i][k] == 'P') {
				for (int j = 0; j < 4; j++) {
					int x = i + dx[j];
					int y = k + dy[j];
					if (x >= 0 && x < N && y >= 0 && y < N) {
						if (place[x][y] == 'P') {
							return false;
						}
						else if (place[x][y] == 'O') {
							for (int f = 0; f < 4; f++) {
								int xx = x + dx[f];
								int yy = y + dy[f];
								if (xx >= 0 && xx < N && yy >= 0 && yy < N) {
									if (xx == x && yy == y) continue;
									if (xx == i && yy == k) continue;
									if (place[xx][yy] == 'P') {
										return false;
									}
								}
							}
						}
					}
				}
			}
		}
	}
	return true;
}

vector<int> solution(vector<vector<string>> places) {
	vector<int> answer;
	for (vector<string> place : places) {
		answer.push_back(check(place));
	}
	return answer;
}
